src/block-gen/sequencer_rollup.py

```diff
@@ -0,0 +1,85 @@
+import pickle
+from typing import List, NewType
+from block_gen import Block, Transaction, Event
+
+# A more complex rollup that allows for privilaged block
+# production. This is achieved by allowing a "sequencer"
+# to squeeze blocks into historical L1 blocks after the fact.
+
+# This expands on the "simplest possible rollup" by, in addition
+# to blocks being generated based on the first event, the sequencer
+# may submit transactions in later blocks which contain many L2 blocks.
+
+#################### Types ####################
+
+class SequencerBlock(Block):
+    # An optional field which specifies the epoch that
+    # this sequencer block is intended for.
+    target_epoch: int
+
+SequencerBatch = NewType("SequencerBatch", List[SequencerBlock])
+
+#################### Functions ####################
+
+# Generate the full rullup chain
+def generate_rollup_chain(l1_chain: List[Block], sequencer_timeout) -> List[Block]:
+    l2_chain: List[Block] = []
+    for i in range(0, len(l1_chain) - sequencer_timeout):
+        root_block = l1_chain[i]
+        subsequent_blocks = l1_chain[i+1:i+1+sequencer_timeout]
+        l2_blocks = generate_rollup_epoch(root_block, subsequent_blocks, sequencer_timeout)
+        l2_chain += l2_blocks
+    return l2_chain
+
+# Generate a single epoch of the rollup chain. There is 1 epoch for every L1 block.
+# Epochs have 1 deposit block and can have variable numbers of sequencer blocks.
+# In the worst case you must wait until the `sequencer_timeout` to determine an
+# epoch's blocks.
+def generate_rollup_epoch(
+            root_block: Block,
+            subsequent_blocks: List[Block],
+            sequencer_timeout) -> List[Block]:
+    assert len(subsequent_blocks) >= sequencer_timeout, \
+        "Cannot determine epoch blocks until sequencer timeout has passed"
+    deposit_block = generate_deposit_block(root_block)
+    l2_chain: List[Block] = [deposit_block]
+    # Determine all sequencer blocks
+    last_target_epoch = 0
+    for block in subsequent_blocks:
+        batch: SequencerBatch = extract_batch(block)
+        if batch == None:
+            continue
+        for seq_block in batch:
+            # Update the last_target_epoch
+            if (seq_block["target_epoch"] > last_target_epoch and
+                seq_block["target_epoch"] < block["block_number"]):
+                last_target_epoch = seq_block["target_epoch"]
+            # Ignore the block if it is targeting the wrong epoch
+            if last_target_epoch != root_block["block_number"]:
```

Comment:
Oh whoops good point. I think that that modification can be made. The reason why I kept around `last_target_epoch` is just because I was thinking of the super optimized case where the sequencer doesn't duplicate which `target_epoch` each block is for, and instead when you're generating the blocks it is assumed that if your block omits the `target_epoch` then it defaults to `last_target_epoch`